text/000-sfc-and-template-import-primitives.md

```diff
@@ -0,0 +1,875 @@
+- Start Date: 2019-02-22
+- Relevant Team(s): Ember.js
+- RFC PR: https://github.com/emberjs/rfcs/pull/454
+- Tracking: (leave this empty)
+
+# SFC & Template Import Primitives
+
+## Summary
+
+Expose low-level primitives for embedding templates in JavaScript and
+associating templates with component classes.
+
+These primitives unlock experimentation, allowing addons to provide
+highly-requested features (such as single-file components) via stable, public
+API.
+
+## Motivation
+
+This proposal is intended to unlock experimentation around two
+highly-requested features:
+
+1. Single-file components.
+2. Module imports in component templates.
+
+Although exploring these features is the primary motivation, an important
+benefit of stabilizing low-level APIs is that they enable the Ember community
+to experiment with new, unexpected ideas.
+
+### Single-File Components
+
+In Ember components today, JavaScript code lives in a `.js` file and template
+code lives in a separate `.hbs` file. Juggling between these two files adds
+friction to the developer experience.
```

Comment:
> Juggling between these two files adds friction to the developer experience.

Can we word this to be more relative? It is not true for everyone.

Reply:
I usually split my editor window to show the JS on the top and HBS on the bottom. I don't experience this _Juggling_ problem. Is it possible that the tools a developer uses contributes more to the problem then the file structure itself?

Reply:
If both the JS and HBS were to be in one file I'd have to split the file anyway and it would make the hot swap more difficult as I'd have to make two parts of a file and jump between them. Not only that but having two syntaxs and linting rules and compilers in one file is a massive undertaking in tooling and comprehension. I like that I can see a `.js` file and know it is JavaScript. I don't have to worry it is a Frankenstein experiment of non-JavaScript. I can `ember-template-lint *.hbs` and not have to also conciser it needs to be integrated into eslint. Doing any kind of find  find or `git grep call-site -- *.hbs` would compound the problem. I could go on but I hope my point is made. JSX is a fanatical nightmare in maintenance, usability, and comprehension. It may look nice for a really small module but I'd ask since when have all your components been that small? Real world apps are huge and complicated and we as devs need all the help we can get making sense of it all--computer tools too.

Reply:
I think @MelSumner's note on this is probably the most apt: developer experiences around this are *mixed*. I have lots of friends who simply don't experience the kind of "nightmare in maintenance, usability, and comprehension" that @sukima does – and I absolutely believe *both* reports.

These kinds of things tend to be very specific to individual tastes (in much the same way that many people love each of Python and Ruby but very few people love both), and they are also profoundly impacted by the patterns used in specific apps and teams.

I also think it's worth remembering that the point here is to *unlock different approaches*, which may allow us to find something that works *better* than today's story *or* the alternatives currently used in other frameworks. Regardless, Ember will certainly continue to have a *good default*.

By way of analogy: we have first-tier supported experience with Ember Data—but don't in the least discourage people from using Apollo.

Reply:
@MelSumner I would argue that this is true for everyone that only has one screen or is working from a laptop (outside, at a coffee shop, or traveling).
Grouping css, html & js together would significantly reduce the amount of files I have opened. My screen would still need to be split to display either the child or parent component

Reply:
One analogy that comes to mind here is _unnecessary abstractions_. If you've ever traced a method call through 2 or 3 or 10 other methods that are all a single-line and thought to yourself "y tho?" you know what I'm talking about.

When a component could fit in its entirety (js + template) on a single (laptop) screen, opening two files, I find myself thinking something similar: "y tho?" The more you split functionality among different files (don't forget CSS!) the more you have to go hunting to get a complete picture of said functionality. And the more you risk inadvertently breaking something when you make changes. Yes, there's a bit of a learning curve here (iykyk) and once you get over the hump, it's not as much of a pain point. But there are certainly circumstances where it's a tool I'd reach for were it available.

@MelSumner I think @chriskrycho's comment is excellent wording:
> the point here is to unlock different approaches, which may allow us to find something that works better than today's story or the alternatives currently used in other frameworks

One use-case that comes to mind is platforms. If I'm building a platform (RIP Boundless Labs 😢), it's more ergonomic (IMO) to ask 3rd party developers to submit components as a single file.

Also, FWIW @fredericrous I am personally one of the rare(?) laptop workers, regardless of where I'm working from.